File: java-algorithm/src/main/java/com/leetcode/Problem847.java
package com.leetcode;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * 847. Shortest Path Visiting All Nodes
 * https://leetcode.com/problems/shortest-path-visiting-all-nodes/
 */
public class Problem847 {
    private int[][] g;
    private boolean[] visited;
    private int ans = Integer.MAX_VALUE;

    public int shortestPathLength(int[][] graph) {
        this.g = graph;
        this.visited = new boolean[graph.length];
        Deque<Integer> queue = new ArrayDeque<>();
        for (int i = 0; i < graph.length; i++) {
            queue.add(i);
        }
        while (queue.size() > 0) {
            int n = queue.size();
            for (int i = 0; i < n; i++) {
                int pos = queue.pollFirst();
                for (int next : graph[pos]) {

                }
            }
        }
        return ans;
    }
}

File: python-algo/leetcode/problem_847.py
"""847. Shortest Path Visiting All Nodes
https://leetcode.com/problems/shortest-path-visiting-all-nodes/
"""
import collections
from typing import List


class Solution:
    def shortest_path_length(self, graph: List[List[int]]) -> int:
        n = len(graph)
        store = collections.deque((i, 1 << i, 0) for i in range(n))
        visited = {(i, 1 << i) for i in range(n)}
        ans = 0
        while store:
            u, bits, steps = store.popleft()
            if bits == (1 << n) - 1:
                return steps
            for v in graph[u]:
                mask = bits | (1 << v)
                if (v, mask) not in visited:
                    visited.add((v, mask))
                    store.append((v, mask, steps + 1))
        return ans
